### ISSUE TITLE
Inlay hints appearing when they should not

### ISSUE THREAD

**Issue by FMorschel** [microsoft/vscode#230498]
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.94.0
- OS Version: Windows 11 Pro

Steps to Reproduce:

1. Have a project with dart and `Dart-code/Dart-code` (https://marketplace.visualstudio.com/items?itemName=Dart-Code.dart-code)
2. Open the project:

https://github.com/user-attachments/assets/67c595d0-ecdd-45e6-ad32-e2bd6b9b8c56

On previous versions this would not appear unless holding down the keys.

---

@DanTup

**Comment by jrieken** [comment]
Trying this with our build in TS support, the settings below, and everything works as expected

```
  "editor.inlayHints.enabled": "on",
  "typescript.inlayHints.parameterNames.enabled": "all",
  "[typescript]": {
    "editor.inlayHints.enabled": "offUnlessPressed",
  },
```

**Comment by FMorschel** [comment]
This only happens when opening a new Dart project with a previously open `.dart` file. If I open VSCode with a Dart project or `Open Folder` with a Dart project this happens.

If we open other files it goes away (for every split editor).

Not sure what else is missing here. Are there any logs I can try and give you here that would help investigate?

**Comment by jrieken** [comment]
Sorry, but I don't have the capacity to try this with dart. Also it should make a difference because all extension provided data is treated equal. Also there is no persisted caching that would explain the "previously open .dart file" portion. 

Unsure when/how the dart setting is defined? Do you have this in your settings json (statically) or is this some dynamic value that is set programmatically (and late)?

**Comment by FMorschel** [comment]
As I understand that dart setting is the default for .dart files as I showed but it's not something I've set on any settings JSON. The extension author @DanTup also has some problems to repro this but there have been some questions about this (as you can see for the above links to this issue and also on Discord for Flutter/Dart). 

That's why I asked about any logs that I could provide to help identifying this issue.

**Comment by DanTup** [comment]
> Do you have this in your settings json (statically) or is this some dynamic value that is set programmatically (and late)?

There's nothing changing this setting dynamically from Dart's perspective - it's just set in `configurationDefaults`. I've also not been able to reproduce it.

My feeling is that maybe there is some race between VS Code restoring editors when it starts up, and reading the extensions manifest to see the `configurationDefaults`? If so, maybe changes that come from `configurationDefaults` are not being applied to existing open editors?

**Comment by FMorschel** [comment]
I found this behaviour on a Galaxy Book Pro.

Tested with both the Windows version on the issue description and Pop!_OS 22.04 and both did the same thing. At work there was another colleague with the same problem, once I'm back at the office I'll see if I can find out what  machine he is using so you can try and repro.

**Comment by FMorschel** [comment]
As an update, the other machine is an IdeaPad 3 15ALC6.

**Comment by RedCMD** [comment]
> it's just set in `configurationDefaults`

@DanTup VSCode doesn't load `configurationDefaults` on reload correctly
https://github.com/microsoft/vscode/issues/231224

@jrieken VSCode isn't setting the setting correctly. plz fix
`"editor.inlayHints.enabled": "offUnlessPressed"`
https://github.com/Dart-Code/Dart-Code/blob/61567e148a52f97d80f809ec9a0313f31946b144/package.json#L1650

**Comment by bpasero** [comment]
Fwiw https://github.com/microsoft/vscode/issues/231224 should be fixed in todays insider build.

**Comment by FMorschel** [comment]
Just downloaded the latest build and the previous one. I can confirm this is WAI now!

**Comment by FMorschel** [comment]
I'm not from the team, but the stable releases usually happen at the start of every month. Not sure about the exact day to inform you, @Rexios80.

**Comment by Rexios80** [comment]
Somehow my inlay hints settings was "on" instead of "offUnlessPressed"... Does the Dart extension override this behavior? I changed the setting to "offUnlessPressed" and the issues seems to have gone away.

**Comment by FMorschel** [comment]
> Does the Dart extension override this behavior?

Yes, it does. On the issue description, I've added a video that shows on the settings menu that this is overridden somewhere else. But it only overrides that for `.dart` files. Everything else will be left as you edited/the default values (which in this case is "on").

**Comment by DanTup** [comment]
I think I just discovered why I was never able to reproduce this... In my User Settings, I had:

```
"[dart]": {
		"editor.inlayHints.enabled": "offUnlessPressed"
	},
```

So I was getting the expected behaviour even without `configurationDefaults` not working correctly 🤦‍♂️ I removed that, and now I am constantly seeing files that are restored when I open projects incorrectly showing the hints. Glad there's a fix on the way!
